Fix normalising smooth cubic Bézier curves without a previous curve

Eschon opened the pull request:
Fixes #11 

choose first control point for smooth cubic bézier curves based on the previous command

if the previous command was another bézier curve use the last control point reflected over the current point otherwise use the current point

Commit:
choose first control point for smooth cubic bézier curves based on the previous command
if the previous command was another bézier curve use the last control point reflected over the current point otherwise use the current point

Pomax commented on reverse.js:
```diff
@@ -34,6 +34,7 @@
         instruction,
         op,
         lop,
+        preop,
```
let's call this `prevop` so that it's clearer what its role is.

Eschon replied:
Ok, done

Pomax commented on reverse.js:
```diff
@@ -194,9 +195,16 @@
-          // reflect previous cx2/cy2 over x/y
-          cx = x + (x-cx2);
-          cy = y + (y-cy2);
+          if (["s", "c"].includes(preop)) {
+            // reflect previous cx2/cy2 over x/y
+            cx = x + (x-cx2);
+            cy = y + (y-cy2);
+          }
+          else {
+            cx = x;
+            cy = y;
+          }
+
```
small tweak, but note that in both cases we assign x/y:
```suggestion
          cx = x;
          cy = y;
          if (["s", "c"].includes(prevop)) {
            cx += (x-cx2);
            cy += (y-cy2);
          }
```

Pomax commented on test.js:
```diff
@@ -75,3 +75,17 @@ prenormalized = "M 0 0 L 100 0 L 100 100 L 20 100 Z";
+prenormalized = "M 110 90 C 130 90 125 10 150 10 C 175 10 170 90 190 90";
+normalized = normalize(path);
+if (normalized !== prenormalized) fail(`smooth cubic Bézier with previous Bézier (normalize): ${normalized}`);
+if (prenormalized !== reverse(reverse(path))) fail('smooth cubic Bézier with previous Bézier');
```
a silly github thing, but this is missing a newline at the end.

Eschon replied:
My bad, added one

Eschon committed:
always assign x/y and reflect last control point in case of previous bézier curve

Co-authored-by: Pomax <pomax@nihongoresources.com>

Commit:
rename preop to prevop

Commit:
add newline at the end of test.js

Pomax commented on reverse.js:
```diff
@@ -195,16 +195,12 @@
-            cy = y + (y-cy2);
+          cx = x;
+          cy = y;
+          if (["s", "c"].includes(prevop)) {
```
```suggestion
          if (["s", "c"].indexOf(prevop) > -1) {
```

Pomax committed:
Update reverse.js

forgot this library was still "universal" JS, meaning we can't rely on the fairly recent `includes` function

Pomax commented:
@Eschon one more thought: this probably also applies to the `t` operator for quadratic curves, would it be too much to ask you to add the same changes there with a quadratic test case?

Commit:
adapt the fix for cubic Bézier curves for quadratic Bézier curves and add test cases

Eschon commented:
@Pomax sure no problem. I've adapted the fix for quadratic curves

Pomax commented on reverse.js:
```diff
@@ -157,9 +157,16 @@
-          // reflect previous cx/cy over x/y
-          cx = x + (x-cx);
-          cy = y + (y-cy);
+          if (["t", "q"].indexOf(prevop) > -1) {
+            // reflect previous cx/cy over x/y
+            cx = x + (x-cx);
+            cy = y + (y-cy);
+          }
+          else {
+            cx = x;
+            cy = y;
+            prevop = lop;
+          }
```
Let's use the same code as for s/c though (universal x/y assignment with conditional increment)

Eschon replied:
Doesn't work in that case. With s/c we reflect cx2/cy2 here we have cx/cy. If we assign them first it breaks

Pomax replied:
d'uh, yes of course.

Pomax commented on test.js:
```diff
@@ -89,3 +89,17 @@ prenormalized = "M 110 90 C 130 90 125 10 150 10 C 175 10 170 90 190 90";
+prenormalized = "M 10 50 Q 25 25 40 50 Q 55 75 70 50 Q 85 25 100 50 Q 115 75 130 50 Q 145 25 160 50 Q 175 75 190 50";
+normalized = normalize(path);
+if (normalized !== prenormalized) fail(`quadratic Bézier with previous Bézier (normalize): ${normalized}`);
+if (prenormalized !== reverse(reverse(path))) fail('quadratic Bézier with previous Bézier');
```
really wish github just ignored this =D

Eschon replied:
my bad, fixed it and also fixed my editor config to add it by default  😅

Commit:
add missing newline at the end of test.js

Pomax commented on reverse.js:
```diff
@@ -156,9 +157,16 @@
+          else {
+            cx = x;
+            cy = y;
+            prevop = lop;
```
We probably want to move this assignment out of the `else`?

Eschon replied:
You're right. Functionally it doesn't really make a difference, I just put it there to make it work.
But it's cleaner to move it out of there

Commit:
move prevop assignment out of the else block for normalizing the "t" command

Pomax reviewed (approved):
Awesome, let's get this in!

Pomax commented:
v1.5.0 released
